InformationScripting/src/graph/Graph.h

```diff
@@ -0,0 +1,129 @@
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#pragma once
+
+#include "../informationscripting_api.h"
+
+namespace InformationScripting {
+
+class InformationEdge;
+class InformationNode;
+
+class Graph
+{
+	public:
+		~Graph();
+		/**
+		 * Each information source can register 1 NodeHash function. The function takes a node as input,
+		 * and should return a pair of the hash value and a bool if the function could succesfully hash the value.
+		 *
+		 * The second parameter is useful as, e.g. a node from the tag information source can not be hashed by the
+		 * ast information source.
+		 */
+		using NodeHash = std::function<QPair<std::size_t, bool>(const InformationNode*)>;
+		using NodeCondition = std::function<bool(const InformationNode*)>;
+		using EdgeCondition = std::function<bool(const InformationEdge*)>;
+
+		static inline void registerNodeHash(NodeHash check);
+
+		/**
+		 * Adds the \a node to the graph. Only the returned node should be used after this method is called.
+		 * (It might be that the graph already contains the same node, thus the properties will be merged)
+		 *
+		 * Note: Takes ownership on \a node.
+		 */
+		InformationNode* add(InformationNode* node);
+
+		/**
+		 * Creates a directed Edge from \a from to \a to with the name \a name.
+		 *
+		 * Returns the directed connection between \a from and \a to.
+		 *
+		 * If there is already a directed edge with the same \a name between \a from and \a to,
+		 * the existing edge is returned with the count property increased.
+		 *
+		 * Note: The graph owns the returned edge.
+		 */
+		InformationEdge* addDirectedEdge(InformationNode* from, InformationNode* to, const QString& name);
+
+		/**
+		 * Connects \a a to \a b with an indirected edge with the name \a name.
+		 *
+		 * Returns the undirected connection between \a a and \a b.
+		 *
+		 * If there is already a undirected edge with the same \a name between \a a and \a b,
+		 * the existing edge is returned with the count property increased.
+		 *
+		 * Note: The graph owns the returned edge.
+		 */
+		InformationEdge* addEdge(InformationNode* a, InformationNode* b, const QString& name);
+
+
+		/**
+		 * Removes the \a node and all edges to it.
+		 */
+		void remove(InformationNode* node);
+
+		/**
+		 * Removes all nodes in the list \a nodes and all edges to the nodes.
+		 */
+		void remove(QList<InformationNode*> nodes);
+
+		QList<InformationNode*> nodes() const;
+
+		/**
+		 * Returns all nodes for which the NodeCondition \a holds holds.
+		 */
+		QList<InformationNode*> nodesForWhich(NodeCondition holds) const;
+
+		/**
+		 * Returns all edges for which the EdgeCondition \a holds holds.
+		 */
+		QList<InformationEdge*> edgesFowWhich(EdgeCondition holds) const;
```

Comment:
Same as above. I assume that at some point we will also need `edges()` so just go ahead and make a common `edges(condition = {})`